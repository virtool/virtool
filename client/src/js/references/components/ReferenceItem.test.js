import ReferenceItemContainer, {
    ReferenceHeader,
    ReferenceMetadata,
    getOrigin,
    ReferenceItem
} from "./ReferenceItem";
import { MemoryRouter } from "react-router";
import { push } from "react-router-redux";

describe("<ReferenceItem />", () => {
    const initialState = {
        processes: {
            documents: [{ id: "123abc", progress: 1, step: "test" }]
        }
    };
    const store = mockStore(initialState);
    let props;
    let wrapper;

    beforeEach(() => {
        props = {
            process: { id: "123abc" },
            created_at: "2018-01-01T00:00:00.000000Z",
            data_type: "test-genome",
            id: "test-reference-id",
            name: "Tester",
            organism: "test-virus",
            user: { id: "test-user" }
        };
    });

    it("renders correctly", () => {
        wrapper = shallow(<ReferenceItemContainer store={store} {...props} />).dive();
        expect(wrapper).toMatchSnapshot();
    });

    it("renders base component correctly", () => {
        props = {...props, processes: []};
        wrapper = shallow(<ReferenceItem {...props} />);
        expect(wrapper).toMatchSnapshot();

        props = {...props, processes: [{ id: "123abc", progress: 0.5, step: "test" }]};
        wrapper = shallow(<ReferenceItem {...props} />);
        expect(wrapper).toMatchSnapshot();

        props = {...props, processes: [{ id: "345def", progress: 1, step: "test" }]};
        wrapper = shallow(<ReferenceItem {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders <ReferenceHeader /> subcomponent correctly", () => {
        props = {
            name: props.name,
            createdAt: props.created_at,
            user: props.user.id,
            refId: props.id
        };
        wrapper = shallow(<ReferenceHeader {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("renders <ReferenceMetadata /> subcomponent correctly", () => {
        props = {
            id: props.id,
            data_type: props.data_type,
            organism: props.organism,
            origin: { method: "Imported from", data: { name: "test-import" } },
            latest_build: {
                id: "test-build",
                version: 0,
                created_at: "2018-01-01T00:00:00.000000Z",
                user: props.user
            },
            progress: 100
        };
        wrapper = shallow(<ReferenceMetadata {...props} />);
        expect(wrapper).toMatchSnapshot();

        wrapper.setProps({
            origin: { method: "Cloned from", data: { id: "123clone", name: "test-clone" } },
            organism: "",
            progress: 0
        });
        expect(wrapper).toMatchSnapshot();

        wrapper.setProps({
            origin: { method: "Remotes from", data: { slug: "test-slug" } }
        });
        expect(wrapper).toMatchSnapshot();

        wrapper.setProps({
            origin: { method: "Created", data: "No File" }
        });
        expect(wrapper).toMatchSnapshot();
    });

    it("getOrigin() helper function", () => {
        props = { imported_from: { foo: "bar" } };
        let result = getOrigin(props);
        let expected = { method: "Imported from", data: props.imported_from };
        expect(result).toEqual(expected);

        props = { cloned_from: { hello: "world" } };
        result = getOrigin(props);
        expected = { method: "Cloned from", data: props.cloned_from };
        expect(result).toEqual(expected);

        props = { remotes_from: { test: "example" } };
        result = getOrigin(props);
        expected = { method: "Remotes from", data: props.remotes_from };
        expect(result).toEqual(expected);

        result = getOrigin({});
        expected = { method: "Created", data: "No File" };
        expect(result).toEqual(expected);
    });

    it("Clicking on reference card dispatches router location change", () => {
        const spy = sinon.spy(store, "dispatch");
        expect(spy.called).toBe(false);

        wrapper = mount(
            <MemoryRouter initialEntries={["/refs/"]}>
                <ReferenceItemContainer store={store} {...props} />
            </MemoryRouter>
        );

        let mockEvent = {
            target: { nodeName: "A"}
        };
        wrapper.children().children().children().prop("onClick")(mockEvent);
        expect(spy.called).toBe(false);

        mockEvent = {
            target: { nodeName: "DIV"}
        };
        wrapper.children().children().children().prop("onClick")(mockEvent);
        expect(spy.calledWith(push("/refs/test-reference-id"))).toBe(true);

        spy.restore();
    });

});
