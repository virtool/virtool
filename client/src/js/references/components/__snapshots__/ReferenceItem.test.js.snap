// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReferenceItem /> renders <ReferenceHeader /> subcomponent correctly 1`] = `
<div
  style={
    Object {
      "marginLeft": "5px",
    }
  }
>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <strong>
      Tester
    </strong>
    <Link
      replace={false}
      style={
        Object {
          "float": "right",
        }
      }
      to={
        Object {
          "state": Object {
            "cloneReference": true,
            "newReference": true,
            "refId": "test-reference-id",
          },
        }
      }
    >
      <Icon
        faStyle="fas"
        fixedWidth={false}
        name="clone"
        pullRight={false}
        tip="Clone"
      />
    </Link>
  </Row>
  <Row
    bsClass="row"
    componentClass="div"
  >
    <small>
      Created 
      <RelativeTime
        time="2018-01-01T00:00:00.000000Z"
      />
       by 
      test-user
    </small>
  </Row>
</div>
`;

exports[`<ReferenceItem /> renders <ReferenceMetadata /> subcomponent correctly 1`] = `
<Table
  bordered={true}
  bsClass="table"
  condensed={false}
  hover={false}
  responsive={false}
  striped={false}
>
  <tbody>
    <tr>
      <th>
        Data Type
      </th>
      <td
        className="text-capitalize"
      >
        test-genome
      </td>
    </tr>
    <tr>
      <th>
        Organism
      </th>
      <td
        className="text-capitalize"
      >
        test-virus
      </td>
    </tr>
    <tr>
      <th>
        Imported from
      </th>
      <td>
        test-import
      </td>
    </tr>
    <React.Fragment>
      <tr>
        <th>
          Latest Build
        </th>
        <td>
          <Link
            replace={false}
            to="/refs/test-reference-id/indexes/test-build"
          >
            Index 
            0
          </Link>
          <div
            className="text-muted"
            style={
              Object {
                "fontSize": "12px",
              }
            }
          >
            Created 
            <RelativeTime
              time="2018-01-01T00:00:00.000000Z"
            />
             by 
            test-user
          </div>
        </td>
      </tr>
    </React.Fragment>
  </tbody>
</Table>
`;

exports[`<ReferenceItem /> renders <ReferenceMetadata /> subcomponent correctly 2`] = `
<Table
  bordered={true}
  bsClass="table"
  condensed={false}
  hover={false}
  responsive={false}
  striped={false}
>
  <tbody>
    <tr>
      <th>
        Data Type
      </th>
      <td
        className="text-capitalize"
      >
        test-genome
      </td>
    </tr>
    <tr>
      <th>
        Organism
      </th>
      <td
        className="text-capitalize"
      >
        unknown
      </td>
    </tr>
    <tr>
      <th>
        Cloned from
      </th>
      <td>
        <Link
          replace={false}
          to="/refs/123clone"
        >
          test-clone
        </Link>
      </td>
    </tr>
  </tbody>
</Table>
`;

exports[`<ReferenceItem /> renders <ReferenceMetadata /> subcomponent correctly 3`] = `
<Table
  bordered={true}
  bsClass="table"
  condensed={false}
  hover={false}
  responsive={false}
  striped={false}
>
  <tbody>
    <tr>
      <th>
        Data Type
      </th>
      <td
        className="text-capitalize"
      >
        test-genome
      </td>
    </tr>
    <tr>
      <th>
        Organism
      </th>
      <td
        className="text-capitalize"
      >
        unknown
      </td>
    </tr>
    <tr>
      <th>
        Remotes from
      </th>
      <td>
        <a
          href="https://www.github.com/test-slug"
          rel="noopener noreferrer"
          target="_blank"
        >
          test-slug
        </a>
      </td>
    </tr>
  </tbody>
</Table>
`;

exports[`<ReferenceItem /> renders <ReferenceMetadata /> subcomponent correctly 4`] = `
<Table
  bordered={true}
  bsClass="table"
  condensed={false}
  hover={false}
  responsive={false}
  striped={false}
>
  <tbody>
    <tr>
      <th>
        Data Type
      </th>
      <td
        className="text-capitalize"
      >
        test-genome
      </td>
    </tr>
    <tr>
      <th>
        Organism
      </th>
      <td
        className="text-capitalize"
      >
        unknown
      </td>
    </tr>
    <tr>
      <th>
        Created
      </th>
      <td>
        No File
      </td>
    </tr>
  </tbody>
</Table>
`;

exports[`<ReferenceItem /> renders base component correctly 1`] = `
<Uncontrolled(Panel)
  className="card reference-item"
>
  <PanelHeading
    bsClass="panel"
    componentClass="div"
  >
    <ReferenceHeader
      createdAt="2018-01-01T00:00:00.000000Z"
      name="Tester"
      refId="test-reference-id"
      user="test-user"
    />
  </PanelHeading>
  <ReferenceMetadata
    created_at="2018-01-01T00:00:00.000000Z"
    data_type="test-genome"
    id="test-reference-id"
    name="Tester"
    organism="test-virus"
    origin={
      Object {
        "data": "No File",
        "method": "Created",
      }
    }
    process={
      Object {
        "id": "123abc",
      }
    }
    processes={Array []}
    progress={100}
    user={
      Object {
        "id": "test-user",
      }
    }
  />
  <PanelBody
    bsClass="panel"
    collapsible={false}
    style={
      Object {
        "padding": 0,
        "textAlign": "center",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "small",
          "visibility": "hidden",
        }
      }
    >
      None
    </span>
  </PanelBody>
  <ProgressBar
    affixed={true}
    bsStyle="success"
    now={100}
  />
</Uncontrolled(Panel)>
`;

exports[`<ReferenceItem /> renders base component correctly 2`] = `
<Uncontrolled(Panel)
  className="card reference-item"
>
  <PanelHeading
    bsClass="panel"
    componentClass="div"
  >
    <ReferenceHeader
      createdAt="2018-01-01T00:00:00.000000Z"
      name="Tester"
      refId="test-reference-id"
      user="test-user"
    />
  </PanelHeading>
  <ReferenceMetadata
    created_at="2018-01-01T00:00:00.000000Z"
    data_type="test-genome"
    id="test-reference-id"
    name="Tester"
    organism="test-virus"
    origin={
      Object {
        "data": "No File",
        "method": "Created",
      }
    }
    process={
      Object {
        "id": "123abc",
      }
    }
    processes={
      Array [
        Object {
          "id": "123abc",
          "progress": 0.5,
          "step": "test",
        },
      ]
    }
    progress={50}
    user={
      Object {
        "id": "test-user",
      }
    }
  />
  <PanelBody
    bsClass="panel"
    collapsible={false}
    style={
      Object {
        "padding": 0,
        "textAlign": "center",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "small",
          "visibility": "visible",
        }
      }
    >
      test
    </span>
  </PanelBody>
  <ProgressBar
    affixed={true}
    bsStyle="warning"
    now={50}
  />
</Uncontrolled(Panel)>
`;

exports[`<ReferenceItem /> renders base component correctly 3`] = `
<Uncontrolled(Panel)
  className="card reference-item"
>
  <PanelHeading
    bsClass="panel"
    componentClass="div"
  >
    <ReferenceHeader
      createdAt="2018-01-01T00:00:00.000000Z"
      name="Tester"
      refId="test-reference-id"
      user="test-user"
    />
  </PanelHeading>
  <ReferenceMetadata
    created_at="2018-01-01T00:00:00.000000Z"
    data_type="test-genome"
    id="test-reference-id"
    name="Tester"
    organism="test-virus"
    origin={
      Object {
        "data": "No File",
        "method": "Created",
      }
    }
    process={
      Object {
        "id": "123abc",
      }
    }
    processes={
      Array [
        Object {
          "id": "345def",
          "progress": 1,
          "step": "test",
        },
      ]
    }
    progress={100}
    user={
      Object {
        "id": "test-user",
      }
    }
  />
  <PanelBody
    bsClass="panel"
    collapsible={false}
    style={
      Object {
        "padding": 0,
        "textAlign": "center",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "small",
          "visibility": "hidden",
        }
      }
    >
      None
    </span>
  </PanelBody>
  <ProgressBar
    affixed={true}
    bsStyle="success"
    now={100}
  />
</Uncontrolled(Panel)>
`;

exports[`<ReferenceItem /> renders correctly 1`] = `
<Uncontrolled(Panel)
  className="card reference-item"
  onClick={[Function]}
>
  <PanelHeading
    bsClass="panel"
    componentClass="div"
  >
    <ReferenceHeader
      createdAt="2018-01-01T00:00:00.000000Z"
      name="Tester"
      refId="test-reference-id"
      user="test-user"
    />
  </PanelHeading>
  <ReferenceMetadata
    created_at="2018-01-01T00:00:00.000000Z"
    data_type="test-genome"
    id="test-reference-id"
    name="Tester"
    onClick={[Function]}
    organism="test-virus"
    origin={
      Object {
        "data": "No File",
        "method": "Created",
      }
    }
    process={
      Object {
        "id": "123abc",
      }
    }
    processes={
      Array [
        Object {
          "id": "123abc",
          "progress": 1,
          "step": "test",
        },
      ]
    }
    progress={100}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    user={
      Object {
        "id": "test-user",
      }
    }
  />
  <PanelBody
    bsClass="panel"
    collapsible={false}
    style={
      Object {
        "padding": 0,
        "textAlign": "center",
      }
    }
  >
    <span
      style={
        Object {
          "fontSize": "small",
          "visibility": "hidden",
        }
      }
    >
      test
    </span>
  </PanelBody>
  <ProgressBar
    affixed={true}
    bsStyle="success"
    now={100}
  />
</Uncontrolled(Panel)>
`;
